api/v1alpha3/cluster_webhook.go

```diff
@@ -0,0 +1,73 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha3
+
+import (
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+func (c *Cluster) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(c).
+		Complete()
+}
+
+// +kubebuilder:webhook:verbs=create;update,path=/validate-cluster-x-k8s-io-v1alpha3-cluster,mutating=false,failurePolicy=fail,groups=cluster.x-k8s.io,resources=clusters,versions=v1alpha3,name=validation.cluster.cluster.x-k8s.io
+// +kubebuilder:webhook:verbs=create;update,path=/mutate-cluster-x-k8s-io-v1alpha3-cluster,mutating=true,failurePolicy=fail,groups=cluster.x-k8s.io,resources=clusters,versions=v1alpha3,name=default.cluster.cluster.x-k8s.io
+
+var _ webhook.Defaulter = &Cluster{}
+var _ webhook.Validator = &Cluster{}
+
+func (c *Cluster) Default() {
+	if c.Spec.InfrastructureRef != nil && len(c.Spec.InfrastructureRef.Namespace) == 0 {
+		c.Spec.InfrastructureRef.Namespace = c.Namespace
+	}
+}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (c *Cluster) ValidateCreate() error {
+	return c.validate()
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (c *Cluster) ValidateUpdate(old runtime.Object) error {
+	return c.validate()
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (c *Cluster) ValidateDelete() error {
+	return nil
+}
+
+func (c *Cluster) validate() error {
+	var allErrs field.ErrorList
+	if c.Spec.InfrastructureRef != nil && c.Spec.InfrastructureRef.Namespace != c.Namespace {
+		allErrs = append(
+			allErrs,
+			field.Forbidden(field.NewPath("spec", "infrastructureRef", "namespace"), "expected spec.infrastructureRef.namespace to match namespace"),
```

Comment:
By specifying the field path as the first argument, we can omit it in the message. I also think Invalid makes more sense here.

```suggestion
			field.Invalid(field.NewPath("spec", "infrastructureRef", "namespace"), c.Spec.InfrastructureRef.Namespace, "must match metadata.namespace"),
```

Reply:
I myself was tied between `Forbidden` and `Invalid`.
I picked `Forbidden` over `Invalid` because as per the godoc on the functions, the values are technically correct "as per formatting rules". `Invalid` would be more suited towards malformed values. 
Let me know if y'all still prefer `Invalid`. It still makes sense to the human user reading it 🤷‍♂

Reply:
Rereading the godoc, I'm also 🤷‍♂ on this one. If you read forbidden, though:

> This is used to
// report valid (as per formatting rules) values which would be accepted under
// some conditions, but which are not permitted by current conditions (e.g.
// security policy).

There is only ever one value for the namespace that is acceptable - where it matches the cluster/machine namespace. So I guess it's not really about some (other) conditions vs. current conditions. I think that's why I leaned toward invalid.

I'm mostly interested in the adjustment to the detail message. I'm happy to lgtm on either forbidden or invalid, and we can adjust later if there's enough of a compelling reason to change.